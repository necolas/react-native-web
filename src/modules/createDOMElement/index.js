import '../injectResponderEventPlugin';

import createDOMProps from '../createDOMProps';
import getAccessibilityRole from '../getAccessibilityRole';
import normalizeNativeEvent from '../normalizeNativeEvent';
import React from 'react';
import StyleRegistry from '../../apis/StyleSheet/registry';

const emptyObject = {};

const roleComponents = {
  article: 'article',
  banner: 'header',
  button: 'button',
  complementary: 'aside',
  contentinfo: 'footer',
  form: 'form',
  link: 'a',
  list: 'ul',
  listitem: 'li',
  main: 'main',
  navigation: 'nav',
  region: 'section'
};

const eventHandlerNames = {
  onClick: true,
  onClickCapture: true,
  onMoveShouldSetResponder: true,
  onMoveShouldSetResponderCapture: true,
  onResponderGrant: true,
  onResponderMove: true,
  onResponderReject: true,
  onResponderRelease: true,
  onResponderTerminate: true,
  onResponderTerminationRequest: true,
  onStartShouldSetResponder: true,
  onStartShouldSetResponderCapture: true,
  onTouchCancel: true,
  onTouchCancelCapture: true,
  onTouchEnd: true,
  onTouchEndCapture: true,
  onTouchMove: true,
  onTouchMoveCapture: true,
  onTouchStart: true,
  onTouchStartCapture: true
};

const wrapEventHandler = handler => e => {
  e.nativeEvent = normalizeNativeEvent(e.nativeEvent);
  return handler(e);
};

const getAccessibilityComponent = (props = emptyObject) => {
  const role = getAccessibilityRole(props);
  if (role === 'heading') {
    const level = props['aria-level'] || 1;
    return `h${level}`;
  }
  return roleComponents[role];
};

const createDOMElement = (component, rnProps) => {
  // use equivalent platform elements where possible
  const accessibilityComponent = getAccessibilityComponent(rnProps);
  const Component = accessibilityComponent || component;

  const domProps = createDOMProps(rnProps, style => StyleRegistry.resolve(style));

  // normalize DOM events to match React Native events
  // TODO: move this out of the render path
  for (const prop in domProps) {
    if (Object.prototype.hasOwnProperty.call(domProps, prop)) {
      const isEventHandler = typeof prop === 'function' && eventHandlerNames[prop];
      if (isEventHandler) {
        domProps[prop] = wrapEventHandler(prop);
      }
    }
  }

  return <Component {...domProps} />;
};

module.exports = createDOMElement;
