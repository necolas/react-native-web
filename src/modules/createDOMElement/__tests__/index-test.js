/* eslint-env jasmine, jest */

import createDOMElement from '..';
import renderer from 'react-test-renderer';

describe('modules/createDOMElement', () => {
  test('renders correct DOM element', () => {
    let component = renderer.create(createDOMElement('span'));
    expect(component.toJSON()).toMatchSnapshot();
    component = renderer.create(createDOMElement('main'));
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('prop "accessibilityLabel"', () => {
    const accessibilityLabel = 'accessibilityLabel';
    const component = renderer.create(createDOMElement('span', { accessibilityLabel }));
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('prop "accessibilityLiveRegion"', () => {
    const accessibilityLiveRegion = 'polite';
    const component = renderer.create(createDOMElement('span', { accessibilityLiveRegion }));
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('prop "accessibilityRole"', () => {
    test('roles', () => {
      const component = renderer.create(createDOMElement('span', { accessibilityRole: 'banner' }));
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('button', () => {
      const component = renderer.create(createDOMElement('span', { accessibilityRole: 'button' }));
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('link and target="_blank"', () => {
      const component = renderer.create(createDOMElement('span', { accessibilityRole: 'link', target: '_blank' }));
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  test('prop "accessible"', () => {
    // accessible (implicit)
    let component = renderer.create(createDOMElement('span', {}));
    expect(component.toJSON()).toMatchSnapshot();
    // accessible (explicit)
    component = renderer.create(createDOMElement('span', { accessible: true }));
    expect(component.toJSON()).toMatchSnapshot();
    // not accessible
    component = renderer.create(createDOMElement('span', { accessible: false }));
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('prop "testID"', () => {
    const component = renderer.create(createDOMElement('span', { testID: 'Example.testID' }));
    expect(component.toJSON()).toMatchSnapshot();
  });
});
