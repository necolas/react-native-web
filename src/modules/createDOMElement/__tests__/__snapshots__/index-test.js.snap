// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/createDOMElement prop "accessibilityLabel" 1`] = `
<span
  aria-label="accessibilityLabel"
/>
`;

exports[`modules/createDOMElement prop "accessibilityLiveRegion" 1`] = `
<span
  aria-live="off"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" button 1`] = `
<button
  role="button"
  type="button"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityComponentType 1`] = `
<button
  role="button"
  type="button"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityComponentType 2`] = `
<a
  role="link"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityTraits 1`] = `
<button
  role="button"
  type="button"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityTraits 2`] = `
<a
  role="link"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" headings 1`] = `
<h1
  role="heading"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" headings 2`] = `
<h3
  aria-level="3"
  role="heading"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" link and target="_blank" 1`] = `
<a
  rel=" noopener noreferrer"
  role="link"
  target="_blank"
/>
`;

exports[`modules/createDOMElement prop "accessibilityRole" roles 1`] = `
<header
  role="banner"
/>
`;

exports[`modules/createDOMElement prop "accessible" 1`] = `<span />`;

exports[`modules/createDOMElement prop "accessible" 2`] = `<span />`;

exports[`modules/createDOMElement prop "accessible" 3`] = `
<span
  aria-hidden={true}
/>
`;

exports[`modules/createDOMElement prop "testID" 1`] = `
<span
  data-testid="Example.testID"
/>
`;

exports[`modules/createDOMElement renders correct DOM element 1`] = `<span />`;

exports[`modules/createDOMElement renders correct DOM element 2`] = `<main />`;
